cardinal/search/filtercomponent.go

```diff
@@ -0,0 +1,103 @@
+package search
+
+import (
+	"github.com/rotisserie/eris"
+
+	"pkg.world.dev/world-engine/cardinal/types"
+	"pkg.world.dev/world-engine/cardinal/types/engine"
+)
+
+// This package involves primitives for search.
+// It involves creating and combining primitives that represent
+// filtering properties on components.
+
+type PredicateEvaluator interface {
+	Evaluate(wCtx engine.Context, id types.EntityID) (bool, error)
+}
+
+type componentFilter[T types.Component] struct {
+	FilterFunc func(comp T) bool
+}
+
+type andFilterComponent struct {
+	filterComponents []PredicateEvaluator
+}
+
+type orFilterComponent struct {
+	filterComponents []PredicateEvaluator
+}
+
+type notFilterComponent struct {
+	filterComponent PredicateEvaluator
+}
+
+func FilterFunction[T types.Component](f func(comp T) bool) PredicateEvaluator {
+	return &componentFilter[T]{
+		FilterFunc: f,
+	}
+}
+
+func (afc *andFilterComponent) Evaluate(wCtx engine.Context, id types.EntityID) (bool, error) {
+	var result = true
+	for _, filterComp := range afc.filterComponents {
+		otherResult, err := filterComp.Evaluate(wCtx, id)
+		if err != nil {
+			continue
```

Comment:
This was done with intent. See the comment on top. 

Basically some entities in the search result just won't contain component related to the property filter. In this case the evaluater will return an error rather then a true or a false. For this case the search algorithm just treats the filter as if it never existed.

Reply:
I don't know the exact right thing to do, but it looks inconsistent to have the notFilterComponent return an error and this continue.

Reply:
> I don't know the exact right thing to do, but it looks inconsistent to have the notFilterComponent return an error and this continue.

Yeah, I'll take a look at this.

Reply:
@jerargus Yeah, you're right there's a subtle thing going on here. 

the and, or filters will never return an error which is inconsistent with Not. The issue is subtle. What happens is the and, or filters default to true. There is a case where this shouldn't happen, it's when all evaluations error out. In this case the filter should also return an error. This makes it consistent with the Not filter.

Reply:
So in short this is how it works. 

Any filter that is composed of other filters, will only return an error if ALL internal filters also return an error. Otherwise if only some internal errors return an error, the filter will treat that individual filter as if it never existed.